File: src/trace.ts
import { uuidv7 } from "uuidv7";

/**
 * Generate a new trace id (UUIDv7 hex string with dashes).
 *
 * UUIDv7 is time-ordered: lexical sort matches chronological sort within
 * a single process, with monotonicity guarantees from the underlying lib
 * (`uuidv7` by LiosK — per-process counter bumps sub-millisecond
 * collisions instead of dropping resolution).
 *
 * Mixing UUIDv7 implementations within a process breaks monotonicity (no
 * shared counter). Always import from this module — never construct a
 * trace id from elsewhere.
 */
export function newTraceId(): string {
  return uuidv7();
}

File: python/ai_ops_control_plane/trace.py
"""UUIDv7 trace id — mirror src/trace.ts.

Pinned to `uuid-utils` (Rust-backed, supports `uuid7()` with per-process
monotonic guarantee). Mixing UUIDv7 implementations within a process
breaks monotonicity (no shared counter). Always import from this module.
"""

from __future__ import annotations

from uuid_utils import uuid7


def new_trace_id() -> str:
    """Generate a new trace id (UUIDv7 string, dashed, per-process monotonic)."""
    return str(uuid7())
